File: 12.py
import random

# Список из 7 разных цветов, каждый цвет представлен символом
colors = ['R', 'G', 'B', 'Y', 'P', 'O', 'V']  # Красный, Зеленый, Синий, Желтый, Фиолетовый, Оранжевый, Белый

# Направления для поиска соседей: вверх, вниз, влево, вправо
directions = [(-1, 0), (1, 0), (0, -1), (0, 1)]  # Поиск по 4 сторонам

# Функция для поиска области одинакового цвета с помощью DFS
def dfs(matrix, visited, x, y, color, N, M):
    stack = [(x, y)]
    area = []
    while stack:
        cx, cy = stack.pop()
        if 0 <= cx < N and 0 <= cy < M and not visited[cx][cy] and matrix[cx][cy] == color:
            visited[cx][cy] = True
            area.append((cx, cy))
            # Проверяем все соседние клетки
            for dx, dy in directions:
                nx, ny = cx + dx, cy + dy
                if 0 <= nx < N and 0 <= ny < M and not visited[nx][ny] and matrix[nx][ny] == color:
                    stack.append((nx, ny))
    return area

# Основная функция
def main():
    # Ввод размеров матрицы
    N = int(input("Введите количество строк N: "))
    M = int(input("Введите количество столбцов M: "))

    # Генерация матрицы случайных цветов
    matrix = [[random.choice(colors) for _ in range(M)] for _ in range(N)]

    # Вывод сгенерированной матрицы
    print("\nСгенерированная матрица:")
    for row in matrix:
        print(' '.join(row))

    # Массив для отслеживания посещенных клеток
    visited = [[False] * M for _ in range(N)]

    # Переменные для хранения наибольшей области
    max_area = []
    max_area_size = 0

    # Поиск наибольшей области одинакового цвета
    for i in range(N):
        for j in range(M):
            if not visited[i][j]:
                area = dfs(matrix, visited, i, j, matrix[i][j], N, M)
                if len(area) > max_area_size:
                    max_area = area
                    max_area_size = len(area)

    # Вывод наибольшей области
    print(f"\nНаибольшая область одинакового цвета (размер: {max_area_size}):")
    for x, y in max_area:
        print(f"({x}, {y}) -> {matrix[x][y]}")

if __name__ == "__main__":
    main()

File: 12.cpp
#include <iostream>
#include <vector>
#include <stack>
#include <cstdlib>
#include <ctime>
#include <utility>

using namespace std;

// Список цветов
const vector<char> colors = {'R', 'G', 'B', 'Y', 'P', 'O', 'V'}; // Красный, Зеленый, Синий, Желтый, Фиолетовый, Оранжевый, Белый

// Направления для поиска соседей: вверх, вниз, влево, вправо
const vector<pair<int, int>> directions = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};

// Функция для поиска области одинакового цвета с помощью DFS
void dfs(const vector<vector<char>>& matrix, vector<vector<bool>>& visited, int x, int y, char color, vector<pair<int, int>>& area) {
    int N = matrix.size();
    int M = matrix[0].size();
    stack<pair<int, int>> toVisit; // Стек для координат

    toVisit.push({x, y});

    while (!toVisit.empty()) {
        auto [cx, cy] = toVisit.top();
        toVisit.pop();

        // Проверяем границы и уже посещенные клетки
        if (cx < 0 || cx >= N || cy < 0 || cy >= M || visited[cx][cy] || matrix[cx][cy] != color) {
            continue;
        }

        visited[cx][cy] = true;
        area.push_back({cx, cy});

        // Добавляем соседей в стек
        for (const auto& [dx, dy] : directions) {
            int nx = cx + dx;
            int ny = cy + dy;
            if (nx >= 0 && nx < N && ny >= 0 && ny < M && !visited[nx][ny] && matrix[nx][ny] == color) {
                toVisit.push({nx, ny});
            }
        }
    }
}

int main() {
    srand(static_cast<unsigned>(time(nullptr)));

    // Ввод размеров матрицы
    int N, M;
    cout << "Введите количество строк N: ";
    cin >> N;
    cout << "Введите количество столбцов M: ";
    cin >> M;

    // Генерация матрицы случайных цветов
    vector<vector<char>> matrix(N, vector<char>(M));
    for (int i = 0; i < N; ++i) {
        for (int j = 0; j < M; ++j) {
            matrix[i][j] = colors[rand() % colors.size()];
        }
    }

    // Вывод сгенерированной матрицы
    cout << "\nСгенерированная матрица:\n";
    for (const auto& row : matrix) {
        for (const auto& cell : row) {
            cout << cell << ' ';
        }
        cout << '\n';
    }

    // Массив для отслеживания посещенных клеток
    vector<vector<bool>> visited(N, vector<bool>(M, false));

    // Список всех областей и их размеров
    vector<vector<pair<int, int>>> all_areas;

    // Поиск всех областей одинакового цвета
    for (int i = 0; i < N; ++i) {
        for (int j = 0; j < M; ++j) {
            if (!visited[i][j]) {
                vector<pair<int, int>> area;
                dfs(matrix, visited, i, j, matrix[i][j], area);
                all_areas.push_back(area);
            }
        }
    }

    // Находим максимальный размер области
    int max_size = 0;
    for (const auto& area : all_areas) {
        if (area.size() > max_size) {
            max_size = area.size();
        }
    }

    // Вывод всех областей максимального размера
    cout << "\nОбласти максимального размера (" << max_size << " клеток):\n";
    for (const auto& area : all_areas) {
        if (area.size() == max_size) {
            cout << "Цвет: " << matrix[area[0].first][area[0].second] 
                 << ", Координаты: ";
            for (const auto& [x, y] : area) {
                cout << "(" << x << ", " << y << ") ";
            }
            cout << '\n';
        }
    }

    return 0;
}
